File: backend/app/schemas/invoices.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class InvoiceLineItemResponse(BaseModel):
    id: UUID
    organization_id: UUID
    invoice_id: UUID
    order_item_id: UUID
    product_id: UUID
    product_sku: str
    product_name: str
    description: str
    quantity: Decimal
    unit_price: Decimal
    discount: Decimal
    tax: Decimal
    line_total: Decimal
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InvoiceCreate(BaseModel):
    order_id: UUID
    due_date: datetime | None = None
    notes: str | None = None


class InvoiceIssue(BaseModel):
    due_date: datetime | None = None
    notes: str | None = None


class InvoicePayment(BaseModel):
    amount: Decimal
    payment_method: str
    reference: str | None = None


class InvoiceVoid(BaseModel):
    notes: str | None = None


class InvoiceResponse(BaseModel):
    id: UUID
    organization_id: UUID
    order_id: UUID
    branch_id: UUID
    customer_id: UUID | None = None
    invoice_number: str
    order_number: str
    branch_name: str
    customer_name: str | None = None
    customer_email: str | None = None
    customer_phone: str | None = None
    status: str
    issued_date: datetime | None = None
    due_date: datetime | None = None
    subtotal: Decimal
    discount: Decimal
    tax: Decimal
    total: Decimal
    amount_paid: Decimal
    notes: str | None = None
    issued_by: UUID | None = None
    created_at: datetime
    updated_at: datetime
    line_items: list[InvoiceLineItemResponse] = []

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/invoices.ts
export type InvoiceLineItem = {
  id: string;
  invoice_id: string;
  product_id: string;
  description: string;
  quantity: string | number;
  unit_price: string | number;
  discount: string | number;
  tax: string | number;
  line_total: string | number;
};

export type Invoice = {
  id: string;
  organization_id: string;
  branch_id: string;
  customer_id: string | null;
  order_id: string;
  invoice_number: string;
  status: "DRAFT" | "ISSUED" | "PAID" | "PARTIAL" | "VOID";
  subtotal: string | number;
  discount_total: string | number;
  tax_total: string | number;
  total: string | number;
  amount_paid: string | number;
  issue_date: string | null;
  due_date: string | null;
  notes: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  line_items?: InvoiceLineItem[];
};

export type InvoiceCreate = {
  order_id: string;
  due_date?: string;
  notes?: string;
};

export type InvoiceIssue = {
  due_date?: string;
  notes?: string;
};

export type InvoicePayment = {
  amount: number;
  payment_method: string;
  reference?: string;
};

export type InvoiceVoid = {
  notes?: string;
};
